src/plugins/BaseObjectSerializer.ts

```diff
@@ -0,0 +1,138 @@
+    if (!(object instanceof Object) || object instanceof String) {
+      return object
+    }
+
+    if (object instanceof DataChunk) {
+      const serialized = await this.SerializeBaseWithClosures(object, [...closures])
+      return new ObjectReference(serialized.id)
+    }
+
+    if (object instanceof Array) {
+      // chunk array into 5000 by default
+      const chunkSize = 5000
+      if (object.length > chunkSize) {
+        let serializedCount = 0
+        const data = new Array<DataChunk>()
+        while (serializedCount < object.length) {
+          const dataChunkCount = Math.min(chunkSize, object.length - serializedCount)
+          data.push(new DataChunk(object.slice(serializedCount, serializedCount + dataChunkCount)))
+          serializedCount += dataChunkCount
+        }
+        return await this.PreserializeObject(data, closures)
+      }
+
+      const convertedList = new Array<any>()
+      for (const element of object) {
+        convertedList.push(await this.PreserializeObject(element, closures))
+      }
+      return convertedList
+    }
+
+    if (object instanceof Object) {
+      return Object.fromEntries(await this.PreserializeEachObjectProperty(object, closures))
+    }
+
+    throw new Error(`Cannot serialize object ${object}`)
+  }
+
+  private async PreserializeEachObjectProperty(
+    o: object,
+    closures: Array<Map<string, number>>
+  ): Promise<Map<string, any>> {
+    const converted = new Map<string, any>()
+
+    for (const key of Object.keys(o)) {
+      const objKey = key as keyof object
+      converted.set(key, await this.PreserializeObject(o[objKey], closures))
+    }
+
+    return converted
+  }
+
+  private async StoreObject(objectId: string, object: Map<string, any>) {
+    for (const transport of this.transports) {
+      await transport.SaveObject(objectId, object)
+    }
+  }
+
+  private SerializeMap(map: Map<string, any>): string {
+    return JSON.stringify(Object.fromEntries(map))
+  }
+
+  private GetId(json: string): string {
+    return MD5(json).toString(enc.Hex)
+  }
+
+  private AddSelfToParentClosures(objectId: string, closureTables: Array<Map<string, number>>) {
+    // only go to closureTable length - 1 because the last closure table belongs to the object with the
+    // provided id
+    const parentClosureTablesCount = closureTables.length - 1
+
+    for (let parentLevel = 0; parentLevel < parentClosureTablesCount; parentLevel++) {
+      const childDepth = parentClosureTablesCount - parentLevel
+      closureTables[parentLevel].set(objectId, childDepth)
+    }
+  }
+}
+
+export class DataChunk implements IBase {
+  public speckle_type: string = 'Speckle.Core.Models.DataChunk'
+  public data: any[]
+  constructor(data: any[] | null) {
+    this.data = data ?? []
+  }
+}
+
+export class ObjectReference implements IBase {
+  public speckle_type: string = 'reference'
+
+  constructor(public referencedId: string) {}
+}
+
+export interface IBase {
+  speckle_type: string
```

Comment:
speckle_type is getter-only on C#, we should make it so here to.. I think the equivalent is `readonly`